# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [1]:
import pandas as pd

url='https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv'
df=pd.read_csv(url)

df.head()

,Unnamed: 0,Customer,State,Customer Lifetime Value,Response,Coverage,Education,Effective To Date,EmploymentStatus,Gender,...,Number of Open Complaints,Number of Policies,Policy Type,Policy,Renew Offer Type,Sales Channel,Total Claim Amount,Vehicle Class,Vehicle Size,Vehicle Type
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2/18/11,Employed,M,...,0.0,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,NaN
1,1,KX64629,California,2228.525238,No,Basic,College,1/18/11,Unemployed,F,...,0.0,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,NaN
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2/10/11,Employed,M,...,0.0,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,1/11/11,Employed,M,...,0.0,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,1/17/11,Medical Leave,F,...,NaN,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,NaN


In [2]:
df.columns= df.columns.str.lower().str.replace(" ","_")

In [3]:
df_filtrado= df[(df['total_claim_amount']< 1000) & (df['response']== 'Yes')]
df_filtrado.head()
df_filtrado.shape

(1399, 26)

In [4]:
df.shape

(10910, 26)

El Dataframe original tiene 10910 clientes y el Dataframe filtrado se ha reducido hasta 1399 clientes que tienen una reclamación baja y han respondido "Yes" en la campaña.

In [5]:
df_filtrado_2= df[df['response']== 'Yes']

In [8]:
df.columns

Index(['unnamed:_0', 'customer', 'state', 'customer_lifetime_value',
       'response', 'coverage', 'education', 'effective_to_date',
       'employmentstatus', 'gender', 'income', 'location_code',
       'marital_status', 'monthly_premium_auto', 'months_since_last_claim',
       'months_since_policy_inception', 'number_of_open_complaints',
       'number_of_policies', 'policy_type', 'policy', 'renew_offer_type',
       'sales_channel', 'total_claim_amount', 'vehicle_class', 'vehicle_size',
       'vehicle_type'],
      dtype='str')

In [9]:
df_filtrado_2.groupby(['policy_type','gender'])[['monthly_premium_auto','customer_lifetime_value']].mean()

monthly_premium_auto  customer_lifetime_value
policy_type    gender                                               
Corporate Auto F                  94.301775              7712.628736
               M                  92.188312              7944.465414
Personal Auto  F                  98.998148              8339.791842
               M                  91.085821              7448.383281
Special Auto   F                  92.314286              7691.584111
               M                  86.343750              8247.088702

En esta última tabla se puede observar que la prima mensual más alta es Personal Auto en género Femenino con un importe de 98.998148. En customer_lifetime_value hay siempre diferencias al alza en las mujeres excepto en Special auto. El segmento más valioso son las mujeres con la póliza Personal auto porque tienen la prima mensual más alta, y generan más valor a largo plazo.

In [14]:
df.groupby('state')['customer'].count()


state
Arizona       1937
California    3552
Nevada         993
Oregon        2909
Washington     888
Name: customer, dtype: int64

In [12]:
conteo_state= df.groupby(['state'])['customer'].count()
conteo_state[conteo_state > 500]

state
Arizona       1937
California    3552
Nevada         993
Oregon        2909
Washington     888
Name: customer, dtype: int64

El resultado da los mismos estados porque toso los estados tenían más de 500 clientes, el más pequeño era Washington con 888.

In [15]:
df.groupby(['education','gender'])['customer_lifetime_value'].agg(['max','min','median'])

max          min       median
education            gender                                       
Bachelor             F       73225.95652  1904.000852  5640.505303
                     M       67907.27050  1898.007675  5548.031892
College              F       61850.18803  1898.683686  5623.611187
                     M       61134.68307  1918.119700  6005.847375
Doctor               F       44856.11397  2395.570000  5332.462694
                     M       32677.34284  2267.604038  5577.669457
High School or Below F       55277.44589  2144.921535  6039.553187
                     M       83325.38119  1940.981221  6286.731006
Master               F       51016.06704  2417.777032  5729.855012
                     M       50568.25912  2272.307310  5579.099207

El grupo que tiene el máximo más alto es High School or Below en la sección masculina. El grupo con el mímimo más alto es Doctor en la sección femenina. Dentro de los mismos niveles educativos hay diferencias entre los sexos siendo en la mayoría de ellos, las mujeres con valores más alto que los hombres a excepción de College y High School or Bellow

In [18]:
df['month'] = df['effective_to_date'].dt.month
df[['effective_to_date', 'month']].head()

,effective_to_date,month
0,2011-02-18,2
1,2011-01-18,1
2,2011-02-10,2
3,2011-01-11,1
4,2011-01-17,1


In [ ]:
df.pivot_table(index ='state',columns='month',values='effective_to_date',aggfunc=)

month,1,2
state,,
Arizona,1008,929
California,1918,1634
Nevada,551,442
Oregon,1565,1344
Washington,463,425


El estado que vendió más pólizas fue California, y el mes que más se vendió fue el mes 1 con 5005

In [20]:
conteo_polizas= df.groupby(['state','month'])['customer'].count()
conteo_polizas

state       month
Arizona     1        1008
            2         929
California  1        1918
            2        1634
Nevada      1         551
            2         442
Oregon      1        1565
            2        1344
Washington  1         463
            2         425
Name: customer, dtype: int64

In [21]:
conteo_polizas.sort_values(ascending=False)

state       month
California  1        1918
            2        1634
Oregon      1        1565
            2        1344
Arizona     1        1008
            2         929
Nevada      1         551
Washington  1         463
Nevada      2         442
Washington  2         425
Name: customer, dtype: int64

In [23]:
top3_estados = conteo_polizas.groupby('state').sum().sort_values(ascending=False).head(3).index
top3_estados

Index(['California', 'Oregon', 'Arizona'], dtype='str', name='state')

In [24]:
conteo_polizas[conteo_polizas.index.get_level_values('state').isin(top3_estados)]

state       month
Arizona     1        1008
            2         929
California  1        1918
            2        1634
Oregon      1        1565
            2        1344
Name: customer, dtype: int64

El estado que lidera más ventas es California en el mes 1 con 1918 pólizas

In [ ]:
df_respuestas= df[(df['response']== 'Yes')]

In [26]:
df_respuestas.groupby('sales_channel')['customer'].count()

sales_channel
Agent          742
Branch         326
Call Center    221
Web            177
Name: customer, dtype: int64

El canal con más respuestas positivas fue Agente, y el que parece que funciona peor es Web